File: c++/pointers/definition_initialisation_and_assignment.cpp
/*
 * Copyright (C) 2019-2026 Katarzyna Krasińska
 * Languagium - https://github.com/katheroine/languagium
 * Licensed under GPL-3.0 - see LICENSE.md
 */
// SPDX-License-Identifier: GPL-3.0-or-later

#include <iostream>

int main() {
  int i1 = 12, i2 = 23;
  int *pi1 = &i1, *pi2 = &i2;

  std::cout
    << "int i1 = 12, i2 = 23;" << std::endl
    << "i1: " << i1 << std::endl
    << "i2: " << i2 << std::endl
    << "int *pi1 = &i1, *pi2 = &i2;" << std::endl
    << "*pi1: " << *pi1 << std::endl
    << "*pi2: " << *pi2 << std::endl
    << std::endl;

  pi1 = &i2; pi2 = &i1;

  std::cout
    << "pi1 = &i2; pi2 = &i1;" << std::endl
    << "*pi1: " << *pi1 << std::endl
    << "*pi2: " << *pi2 << std::endl
    << std::endl;

  *pi1 = 46; *pi2 = 24;

  std::cout
    << "*pi1 = 46; *pi2 = 24;" << std::endl
    << "i1: " << i1 << std::endl
    << "i2: " << i2 << std::endl
    << std::endl;
}

File: pascal/pointers/definition_initialisation_and_assignment.pas
(*
 * Copyright (C) 2019-2026 Katarzyna Krasińska
 * Languagium - https://github.com/katheroine/languagium
 * Licensed under GPL-3.0 - see LICENSE.md
 *)

program PointersDefinitionInitialisationAndAssignment;

var
  i1: integer = 12;
  i2: integer = 23;

  pi1: ^integer = @i1;
  pi2: ^integer = @i2;

begin
  writeln('i1: integer = 12; i2: integer = 23;');
  writeln('i1: ', i1);
  writeln('i2: ', i2);
  writeln('pi1: ^integer = @i1; pi2: ^integer = @i2;');
  writeln('pi1^: ', pi1^);
  writeln('pi2^: ', pi2^);
  writeln();

  pi1 := @i2; pi2 := @i1;

  writeln('pi1 := @i2; pi2 := @i1;');
  writeln('pi1^: ', pi1^);
  writeln('pi2^: ', pi2^);
  writeln();

  pi1^ := 46; pi2^ := 24;

  writeln('pi1^ := 46; pi2^ := 24;');
  writeln('i1: ', i1);
  writeln('i2: ', i2);
  writeln();
end.
